pkg/chain/ethereum/ethutil/mining_waiter_test.go

```diff
@@ -0,0 +1,836 @@
+		if attemptsSoFar == expectedAttempts {
+			chain.receipt = &types.Receipt{}
+		} else {
+			attemptsSoFar++
+		}
+		return createDynamicFeeTransaction(
+			newTransactorOptions.GasFeeCap,
+			newTransactorOptions.GasTipCap,
+		), nil
+	}
+
+	waiter := NewMiningWaiter(chain, config)
+	waiter.ForceMining(
+		originalTransaction,
+		originalTransactorOptions,
+		resubmitFn,
+	)
+
+	resubmissionCount := len(resubmissions)
+	if resubmissionCount != expectedAttempts {
+		t.Fatalf(
+			"expected [%v] resubmission; has: [%v]",
+			expectedAttempts,
+			resubmissionCount,
+		)
+	}
+
+	for index, resubmission := range resubmissions {
+		assertNonceUnchanged(t, resubmission)
+
+		if resubmission.GasLimit != originalTransaction.Gas() {
+			t.Fatalf(
+				"resubmission [%v] gas limit should be the same as in "+
+					"original transaction",
+				index,
+			)
+		}
+
+		if resubmission.GasPrice != nil {
+			t.Fatalf("resubmission [%v] gas price should be nil", index)
+		}
+
+		expectedGasFeeCap := expectedResubmissionParams[index].gasFeeCap
+		if resubmission.GasFeeCap.Cmp(expectedGasFeeCap) != 0 {
+			t.Fatalf(
+				"unexpected resubmission [%v] gas fee cap value\n"+
+					"expected: [%v]\n"+
+					"actual:   [%v]",
+				index,
+				expectedGasFeeCap,
+				resubmission.GasFeeCap,
+			)
+		}
+
+		expectedGasTipCap := expectedResubmissionParams[index].gasTipCap
+		if resubmission.GasTipCap.Cmp(expectedGasTipCap) != 0 {
+			t.Fatalf(
+				"unexpected resubmission [%v]  gas tip cap value\n"+
+					"expected: [%v]\n"+
+					"actual:   [%v]",
+				index,
+				expectedGasTipCap,
+				resubmission.GasTipCap,
+			)
+		}
+	}
+}
+
+func TestForceMining_DynamicFee_MaxAllowedPriceReached(t *testing.T) {
+	originalBaseFee := big.NewInt(10000000000)   // 10 Gwei
+	originalGasTipCap := big.NewInt(4000000000)  // 4 Gwei
+	originalGasFeeCap := big.NewInt(24000000000) // 24 Gwei (2 * baseFee + gasTipCap)
+
+	originalTransaction := createDynamicFeeTransaction(
+		originalGasFeeCap,
+		originalGasTipCap,
+	)
+
+	chain := &mockAdaptedEthereumClientWithReceipt{
+		mockAdaptedEthereumClient: &mockAdaptedEthereumClient{},
+	}
+
+	// Massive increase of base fee to 30 Gwei. This is needed
+	// to exceed the maximum gas fee cap value.
+	chain.blocks = append(chain.blocks, big.NewInt(1))
+	chain.blocksBaseFee = append(
+		chain.blocksBaseFee,
+		new(big.Int).Mul(originalBaseFee, big.NewInt(3)),
+	)
+
+	var resubmissions []*bind.TransactOpts
+
+	resubmitFn := func(
+		newTransactorOptions *bind.TransactOpts,
+	) (*types.Transaction, error) {
+		resubmissions = append(resubmissions, newTransactorOptions)
+		// Not setting mockBackend.receipt, mining takes a very long time.
+		return createDynamicFeeTransaction(
+			newTransactorOptions.GasFeeCap,
+			newTransactorOptions.GasTipCap,
+		), nil
+	}
+
+	waiter := NewMiningWaiter(chain, config)
+	waiter.ForceMining(
+		originalTransaction,
+		originalTransactorOptions,
+		resubmitFn,
+	)
+
+	resubmissionCount := len(resubmissions)
```

Comment:
We could drop a comment that 2*30 + 4.8 > 45 and that's why we resubmit only one time, with 45 fee cap.

Reply:
See https://github.com/keep-network/keep-common/pull/84/commits/f9f2ef97